py/dml/messages.py

```diff
@@ -829,6 +829,18 @@ class ECONDINEACH(DMLError):
     version = "1.4"
     fmt = "conditional 'in each' is not allowed"
 
+class EBADCONDSTMT(DMLError):
+    """
+    `#if` statements in object scope are only allowed to contain
+    certain kinds of declarations: objects, `method`, `session`,
+    `saved`, `#if`, or `error`.
+
+    A special exception is that a `#if` on top scope may contain any
+    kind of statement as long as the `#if` condition doesn't reference
+    any identifiers other than `dml_1_2`, `true`, and `false`.
+    """
+    fmt = "object of type %s not allowed inside `#if`"
```

Comment:
> `%s declarations are not allowed inside #if`

Reply:
Or maybe `%s statements are not ...`? Note that this is a best-effort error message, will e.g. say "extern_typedef statements are ..." which is not optimal but good enough. The important ones are `param`, `import` and `template`, and we get those right at least.

Reply:
How come you prefer `statements` instead of `declarations`? Sure, we have a concept of "object statements", but as you point out, many of the forbidden things doesn't the concept; and we have "declaration" has a collective name for object statements and global declarations (which sometimes is not ideal; for example in our documentation we often say "param declarations" as a collective name for all param object statements, which I _dislike_, but that's what we do. If we say "param declarations not allowed inside `#if`" I don't think people would get confused.)

Reply:
Also, this error message should be tagged `version = "1.4"`. Toplevel if logic can only happen in 1.4 code.

Reply:
_I just realized._ The emission of this error message disagrees with _non_-toplevel ifs. They crop up from invocations of `validate_if_body()` from dmlparse.py; and those use `ECONDP`, `ECONDT` and `ECONDINEACH`. We'll have to address that; replace them all with `EBADCONDSTMT`?

Reply:
I don't see a contradiction here. The messages for non-top-level ifs do not need to mention that top-level ifs have a special case. They are also more specific than top-level ifs, and I don't see a problem with that.  I would probably not oppose unifying the messages, but that can be done in a separate PR.

Reply:
> The messages for non-top-level ifs do not need to mention that top-level ifs have a special case. 

Well, no, I'm not arguing that (we've settled that a time ago.) I'm confused why you're bringing that up; what are you addressing with that statement? Or are you talking about documentation -- "the _documentation_ for non-top-level ifs do not need to mention that top-level ifs have a special case, thus it's a good thing they're separate error classes?" I don't buy **that** argument at all.

> They are also more specific than top-level ifs, and I don't see a problem with that.

Neither do I. What I have a problem with is that your PR **introduces** an inconsistency about what error classes are used for reporting the same kind of error (from the user's perspective.) Because of that, I don't buy the argument that unification can be done in a separate PR.
Thus I'm arguing either you replace `ECONDP`, `ECONDT`, and `ECONDINEACH` all with `EBADCONDSTMT`, or you replace `EBADCONDSTMT` with uses of `ECONDP`, `ECONDT`, and `ECONDINEACH`. Either way, you can implement it by having ast builder pass what forbidden declarations are in use, and then you can either report `ECONDP`/`ECONDT`/`ECONDINEACH` for each of them; or you can pass the sites and declaration kinds to `EBADCONDSTMT` to each be printed via `print_site_message`.

Reply:
> What I have a problem with is that your PR introduces an inconsistency

It's not an inconsistency toward the end user. The same kind of mistake will always give an error, and in some circumstances it will use a somewhat different wording (essentially because there's more error cases to cover). The end result is that the user will read the message and adjust their code, and then they will not see the message again. What's the problem? The error tags are mainly related to our internal testing.

Hm, what we *could* do is to pick a more directed message, like "not allowed inside if, except dml_1_2" if the violating statement is one that only can appear on top level, like a typedef or template. And use a plain ECONDP etc for the other cases. Not sure if that makes anyone happier, though.